asset-transfer-events/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,363 @@
+        client,
+        identity: await newIdentity(),
+        signer: await newSigner(),
+    });
+
+    try {
+
+        // Get a network instance representing the channel where the smart contract is deployed.
+        const network = gateway.getNetwork(channelName);
+
+        // Get the smart contract from the network.
+        const contract = network.getContract(chaincodeName);
+
+        //Start Listening to events.
+        startEventListening(network)
+
+        // Create a new asset on the ledger.
+        await createAsset(contract);
+
+        // Update an existing asset asynchronously.
+        await transferAssetAsync(contract);
+
+        // Get the asset details by assetID.
+        await readAssetByID(contract);
+
+        // Delete asset by assetID.
+        await deleteAssetByID(contract);
+
+        // Update an asset which does not exist.
+        await updateNonExistentAsset(contract)
+
+        console.log('************* BLOCK EVENTS with PRIVATE DATA **************');
+
+        //Generate new assetID
+        assetId = `asset${Date.now()}`;
+
+        // Create a new asset with private data on the ledger.
+        await createAssetPrivate(contract);
+
+        // Update an existing asset with private data asynchronously.
+        await transferAssetAsyncPrivate(contract);
+
+        // Get the asset details along with the private data by assetID.
+        await readAssetByIDPrivate(contract);
+
+    } finally {
+        gateway.close();
+        client.close();
+    }
+}
+
+main().catch(error => console.error('******** FAILED to run the application:', error));
+
+async function newGrpcConnection(): Promise<grpc.Client> {
+    const tlsRootCert = await fs.readFile(tlsCertPath);
+    const tlsCredentials = grpc.credentials.createSsl(tlsRootCert);
+    return new grpc.Client(peerEndpoint, tlsCredentials, {
+        'grpc.ssl_target_name_override': 'peer0.org1.example.com',
+    });
+}
+
+async function newIdentity(): Promise<Identity> {
+    const credentials = await fs.readFile(certPath);
+    return { mspId, credentials };
+}
+
+async function newSigner(): Promise<Signer> {
+    const files = await fs.readdir(keyDirectoryPath);
+    const keyPath = path.resolve(keyDirectoryPath, files[0]);
+    const privateKeyPem = await fs.readFile(keyPath);
+    const privateKey = crypto.createPrivateKey(privateKeyPem);
+    return signers.newPrivateKeySigner(privateKey);
+}
+
+async function startEventListening(network: Network) {
+    console.log('Read chaincode events');
+    const events = await network.getChaincodeEvents(chaincodeName);
+    try {
+        for await (const event of events) {
+            const payload = utf8Decoder.decode(event.payload);
+            console.log(`Received event name: ${event.eventName}, payload: ${payload}, txID: ${event.transactionId}, blockNumber:${event.blockNumber}`);
+        }
+    } finally {
+        // Ensure event iterator is closed when done reading.
+        events.close();
+    }
+}
```

Comment:
This code is problematic as the application just hangs at the end of the main body because this event reading loop is still running. `events.close()` should probably get called to close the event stream at the end of the main application body. Unfortunately that will mean this for-await loop will terminate with an error because the stream it's reading from is cancelled.

There are a few options. If you want to keep this simple (non-terminating) event reading loop then you might have to:

1. Return the `events` variable to the main function so it can call `close()` on it at the end of execution.
2. Handle any exception thrown here gracefully.

To close the events correctly, you could change the `main()` function to include something like:
```typescript
let events: CloseableAsyncIterable<ChaincodeEvent> | undefined;
try {
    events = await startEventListening(network);
} finally {
    events?.close();
```

To handle an error terminating the for-await loop you might check the gRPC status code to see if it was cancelled (by your call to `events.close()`), something like:
```typescript
async function startEventListening(network: Network): Promise<CloseableAsyncIterable<ChaincodeEvent>> {
    console.log('Read chaincode events');
    const events = await network.getChaincodeEvents(chaincodeName);
    readEvents(events); // Don't await -- let this continue in the background
    return events;
}

async function readEvents(events: CloseableAsyncIterable<ChaincodeEvent>): Promise<void> {
    try {
        for await (const event of events) {
            const payload = utf8Decoder.decode(event.payload);
            console.log(`Received event name: ${event.eventName}, payload: ${payload}, txID: ${event.transactionId}, blockNumber:${event.blockNumber}`);
        }
    } catch (error: unknown) {
        if (!(error instanceof GatewayError) || error.code !== grpc.status.CANCELLED) {
            throw error;
        }
    }
}
```

Another alternative might be to get an AsyncIterator from the AsyncIterable with `const iterator = events[Symbol.asyncIterator]();` and then read from the iterator using `await iterator.next()` after submitting a transaction to receive the event you expect to be emitted. The syntax is not very pretty though.

Similar to using the iterator directly as I just described, but without having to dig into the AsyncIterator API, you could have a helper function to read a single event that uses the for-await syntax. Something like:
```typescript
async function readEvent(events: CloseableAsyncIterable<ChaincodeEvent>): Promise<void> {
    for await (const event of events) {
        const payload = utf8Decoder.decode(event.payload);
        console.log(`Received event name: ${event.eventName}, payload: ${payload}, txID: ${event.transactionId}, blockNumber:${event.blockNumber}`);
        break;
    }
}
```